torch/utils/data/dataset.py

```diff
@@ -260,6 +265,132 @@ def __len__(self):
         return len(self.indices)
 
 
+class ChunkDataset(IterableDataset):
+    r"""Dataset which uses hierarchical sampling for efficient data reading.
+
+
+    ``ChunkDataset`` is a stateful dataset that supports hierarchical sampling and
+    efficient reading through chunks. A chunk is selected based on a sampling strategy
+    (first level) and its content is shuffled (second level) before producing
+    a batch or an example.
+
+    In this context, a chunk could be a file, contents of a folder,
+    section of large text file, etc. In distributed training, to eliminate different workers reading
+    uneven amount of data, the chunks need to be roughly the same size.
+    Please consult ChunkDataReader for more details.
+
+    ``ChunkDataset`` extends ``IterableDataset`` because the length of the dataset
+    (or simply the size) is unknown. Only the number of chunks is known for the dataset.
+
+    In a distributed setup, each `DataLoader` worker has an instance of ``ChunkDataset`` that uses
+    ``DistributedChunkSampler`` sampler to select chunks based on their global worker rank.
+    Once a chunk index has been selected, it is passed to ``ChunkDataReader`` that will load
+    data into the ``ChunkDataset`` internal cache. Only then ``ChunkDataset`` returns batches.
+
+    Arguments:
+        chunk_sampler (DistributedSampler or DistributedChunkSampler): Draw indices for chunks.
+            Typically used to split data amongst dataloader workers (first level of sampling)
+        chunk_reader (ChunkDataReader): Specialized reader for a given input type
+        shuffle_cache (bool): Setting `True` trigger shuffling of the internal chunk cache
+            (second level of sampling) (default: `True`)
+    """
+
+    def __init__(self, chunk_sampler, chunk_reader, shuffle_cache=True):
```

Comment:
@cpuhrsch We still need to think about the API design. Not sure it's conventional to have an `Sampler` as the argument in the constructor.

Reply:
Hi @zhangguanheng66 and thanks for the feedback!

Your concern makes total sense. The decision for adding the sampler at the ChunkDataset was due to some factors:

1) When using `IterableDataset`, Pytorch `DataLoader` doesnt use any user-provided sampler. Instead, it creates a hidden `_InfiniteConstantSampler` that always return `None`. With that approach, we dont have support from `DataLoader` to help coordinating dataset fragmentation between workers. It is up to the `IterableDataset`, this is where our `DistributedChunkSampler` played its role. From the `IterableDataset` docstring, we have the following: 

> each worker process will have a different copy of the dataset object, so it is often desired to configure each copy independently to avoid having duplicate data returned from the workers

It also suggests using

> dataset's :meth:`__iter__` method or the :class:`~torch.utils.data.DataLoader` 's :attr:`worker_init_fn` option to modify each copy's behavior

but this methos on their own don't coordination between workers.

2) Also, as you mentioned before, `ChunkDataset` can be used to prevent out-of-memory issues by splitting dataset in several smaller pieces and caching only a few at a time. As shown by the backing article, good randomization is kept and it is very close to randomizing the whole dataset and just returning a batch from it. The advantage on this approach is that different sampling can be implemented and used on `ChunkDataset` without changing the API.

Reply:
I agree with @thiagocrepaldi that a sampler is needed in here. `ChunkDataset` is effectively a transformer on the `chunk_reader` (which should really be a regular dataset and not a dedicated subclass). Also, iterable datasets should never have a `reset` method! At each call to `__iter__` you should create a completely fresh class with a new instance of the sampler (possibly shuffled), and with a new iteration state.

Reply:
I agree we could move the `self._chunk_sampler_iter = iter(self.chunk_sampler)` from reset to `def __iter__`. However, as the `sampler` is needed inside the `ChunkDataset`, we need a way to inject `epoch` into the dataset's sampler (users may wish to set a new seed every epoch). Today this is done through `reset`, but we can rename it to `set_epoch` and move the iterator part to `def __iter__()`

Reply:
Why does this need information about ```epoch```?

Reply:
@cpuhrsch epoch can be used as the seed by every worker to get the same shuffle order.

Reply:
Wouldn't it be easier to pass that order as an order (i.e. as a list or such) before construction?

Reply:
@cpuhrsch That is a good suggestion, but this requires the user know about the number of epochs before hand, which is not a bad thing either. Typically, in training the entire dataset is consumed in an epoch, but for large datasets an epoch could be only a part of the whole dataset (e.g. one few gigabyte file is an epoch and the dataset is many such files). In either case, I think this is a good suggestion and perhaps we can get rid of the reset method.

Reply:
Although it works, it seems a bit odd because none of Pytorch's  `Sampler`s or `Dataset`s know how many epochs they are going to be used for during initialization. That is more of a `DataLoader` domain.

Resetting sampler state after each epoch, on the other hand, is something that we [already do](https://pytorch.org/tutorials/beginner/aws_distributed_training_tutorial.html) during training. 

By renaming `reset` to `set_epoch`, the training loop would change from a typical (before this PR)

```python
for epoch in range(num_epochs):
    train_sampler.set_epoch(epoch)
```
to
```python
for epoch in range(num_epochs):
    train_dataset.set_epoch(epoch)
```

which should look familiar to the user

Reply:
Also, you can use worker_init_fn for DataLoader to set the seed for each worker initially.

Reply:
`worker_init_fn` runs just once, so it wouldnt solve the issue of changing seed for every epoch